Assets/Documentation/man/man3/ClipboardEmergencyShutdown.3
.TH "ClipboardEmergencyShutdown" 3 "Version 0.1" "NPP VR Simulation" \" -*- nroff -*-
.ad l
.nh
.SH NAME
ClipboardEmergencyShutdown \- This class implements logic to trigger the emergency shutdown scenario of the \fBNPPClient\fP class\&.  

.SH SYNOPSIS
.br
.PP
.PP
Inherits MonoBehaviour\&.
.SS "Public Attributes"

.in +1c
.ti -1c
.RI "InputAction \fBactionTrigger\fP"
.br
.RI "
.PP
\fBParameters\fP
.RS 4
\fIactionTrigger\fP is a reference to an InputAction
.RE
.PP
"
.ti -1c
.RI "InputActionAsset \fBclipboardActions\fP"
.br
.RI "
.PP
\fBParameters\fP
.RS 4
\fIclipboardActions\fP is a reference to an InputActionAsset
.RE
.PP
"
.in -1c
.SS "Private Member Functions"

.in +1c
.ti -1c
.RI "void \fBStart\fP ()"
.br
.RI "This method initializes clipboardActions and adds ActionListeners to the clipboard's grab interactable\&. "
.ti -1c
.RI "void \fBOnDestroy\fP ()"
.br
.RI "This method removes the ActionListener, disables the ActionTrigger and destroys the object\&. "
.ti -1c
.RI "void \fBOnSelectEntered\fP (SelectEnterEventArgs args)"
.br
.RI "This method is called when the player interacts with the clipboard\&. "
.ti -1c
.RI "void \fBOnSelectExited\fP (SelectExitEventArgs args)"
.br
.RI "This method is called when the player no longer interacts with the clipboard\&. "
.ti -1c
.RI "void \fBOnActionTriggered\fP (InputAction\&.CallbackContext context)"
.br
.RI "This method is called when the actionTrigger is released, initiating the emergency shutdown scenario on nppClient\&. "
.in -1c
.SS "Private Attributes"

.in +1c
.ti -1c
.RI "\fBNPPClient\fP \fBnppClient\fP"
.br
.RI "
.PP
\fBParameters\fP
.RS 4
\fInppClient\fP is a reference to an instance of the \fBNPPClient\fP class
.RE
.PP
"
.ti -1c
.RI "bool \fBisInteracting\fP = false"
.br
.RI "
.PP
\fBParameters\fP
.RS 4
\fIisInteracting\fP tracks whether the player interacts with the object this script is attached to
.RE
.PP
"
.ti -1c
.RI "UnityEngine\&.XR\&.Interaction\&.Toolkit\&.Interactors\&.XRBaseInteractor \fBinteractor\fP"
.br
.RI "
.PP
\fBParameters\fP
.RS 4
\fIinteractor\fP os a reference to ta XRBaseInteractor
.RE
.PP
"
.in -1c
.SH "Detailed Description"
.PP 
This class implements logic to trigger the emergency shutdown scenario of the \fBNPPClient\fP class\&. 
.SH "Member Function Documentation"
.PP 
.SS "void ClipboardEmergencyShutdown\&.OnActionTriggered (InputAction\&.CallbackContext context)\fR [private]\fP"

.PP
This method is called when the actionTrigger is released, initiating the emergency shutdown scenario on nppClient\&. 
.PP
\fBParameters\fP
.RS 4
\fIcontext\fP passes event specific arguments
.RE
.PP

.SS "void ClipboardEmergencyShutdown\&.OnDestroy ()\fR [private]\fP"

.PP
This method removes the ActionListener, disables the ActionTrigger and destroys the object\&. 
.SS "void ClipboardEmergencyShutdown\&.OnSelectEntered (SelectEnterEventArgs args)\fR [private]\fP"

.PP
This method is called when the player interacts with the clipboard\&. 
.PP
\fBParameters\fP
.RS 4
\fIargs\fP passes event specific arguments at the beginning of the interaction
.RE
.PP

.SS "void ClipboardEmergencyShutdown\&.OnSelectExited (SelectExitEventArgs args)\fR [private]\fP"

.PP
This method is called when the player no longer interacts with the clipboard\&. 
.PP
\fBParameters\fP
.RS 4
\fIargs\fP passes event specific arguments at the end of the interaction
.RE
.PP

.SS "void ClipboardEmergencyShutdown\&.Start ()\fR [private]\fP"

.PP
This method initializes clipboardActions and adds ActionListeners to the clipboard's grab interactable\&. 
.SH "Member Data Documentation"
.PP 
.SS "InputAction ClipboardEmergencyShutdown\&.actionTrigger"

.PP

.PP
\fBParameters\fP
.RS 4
\fIactionTrigger\fP is a reference to an InputAction
.RE
.PP

.SS "InputActionAsset ClipboardEmergencyShutdown\&.clipboardActions"

.PP

.PP
\fBParameters\fP
.RS 4
\fIclipboardActions\fP is a reference to an InputActionAsset
.RE
.PP

.SS "UnityEngine\&.XR\&.Interaction\&.Toolkit\&.Interactors\&.XRBaseInteractor ClipboardEmergencyShutdown\&.interactor\fR [private]\fP"

.PP

.PP
\fBParameters\fP
.RS 4
\fIinteractor\fP os a reference to ta XRBaseInteractor
.RE
.PP

.SS "bool ClipboardEmergencyShutdown\&.isInteracting = false\fR [private]\fP"

.PP

.PP
\fBParameters\fP
.RS 4
\fIisInteracting\fP tracks whether the player interacts with the object this script is attached to
.RE
.PP

.SS "\fBNPPClient\fP ClipboardEmergencyShutdown\&.nppClient\fR [private]\fP"

.PP

.PP
\fBParameters\fP
.RS 4
\fInppClient\fP is a reference to an instance of the \fBNPPClient\fP class
.RE
.PP


.SH "Author"
.PP 
Generated automatically by Doxygen for NPP VR Simulation from the source code\&.


Assets/Skripte/Clipboards/ClipboardEmergencyShutdown.cs
using UnityEngine;
using UnityEngine.XR.Interaction.Toolkit;
using UnityEngine.InputSystem;

/// <summary>
/// This class implements logic to trigger the emergency shutdown scenario of the NPPClient class.
/// </summary>
public class ClipboardEmergencyShutdown : MonoBehaviour
{
    /// <param name="nppClient"> is a reference to an instance of the NPPClient class</param>
    private NPPClient nppClient; 
    /// <param name="actionTrigger">  is a reference to an InputAction</param>
    public InputAction actionTrigger;
    /// <param name="isInteracting"> tracks whether the player interacts with the object this script is attached to</param>
	private bool isInteracting = false;
    /// <param name="interactor"> os a reference to ta XRBaseInteractor</param>
	private UnityEngine.XR.Interaction.Toolkit.Interactors.XRBaseInteractor interactor;
	/// <param name="clipboardActions"> is a reference to an InputActionAsset</param>
	public InputActionAsset clipboardActions;

    /// <summary>
    /// This method initializes clipboardActions and adds ActionListeners to the clipboard's grab interactable.
    /// </summary>
    private void Start()
    {
        // Suche nach der NPPClient-Instanz in der Szene

        nppClient = FindObjectOfType<NPPClient>();
        if (nppClient == null)
        {
            Debug.LogError("NPPClient instance not found in the scene.");
            return;
        }
		
        // Add ActionListeners to the XRGrabInteractable of the clipboard

		var interactable = GetComponent<UnityEngine.XR.Interaction.Toolkit.Interactables.XRGrabInteractable>();
        interactable.selectEntered.AddListener(OnSelectEntered);
        interactable.selectExited.AddListener(OnSelectExited);
		
        // Add and enable ActionTrigger

		var actionMap = clipboardActions.FindActionMap("Clipboard");
		
		if (actionMap == null)
		{
			Debug.LogError("Action Map 'Clipboard' not found in InputActionAsset.");
			return;
		}
		
		actionTrigger = actionMap.FindAction("TriggerClipboardScenario");
		
		if (actionTrigger == null)
		{
			Debug.LogError("Action 'TriggerClipboardScenario' not found in Action Map 'Clipboard'.");
			return;
		} else {
			actionTrigger.Enable();
            actionTrigger.performed += OnActionTriggered;
		}
    }

    /// <summary>
    /// This method removes the ActionListener, disables the ActionTrigger and destroys the object.
    /// </summary>
    private void OnDestroy()
    {
        if (actionTrigger != null)
        {
            actionTrigger.performed -= OnActionTriggered;
            actionTrigger.Disable();
        }
    }
	
    /// <summary>
    /// This method is called when the player interacts with the clipboard.
    /// </summary>
    /// <param name="args"> passes event specific arguments at the beginning of the interaction</param>
	private void OnSelectEntered(SelectEnterEventArgs args)
    {
        isInteracting = true;
        interactor = args.interactorObject as UnityEngine.XR.Interaction.Toolkit.Interactors.XRBaseInteractor;
    }

    /// <summary>
    /// This method is called when the player no longer interacts with the clipboard.
    /// </summary>
    /// <param name="args"> passes event specific arguments at the end of the interaction</param>
    private void OnSelectExited(SelectExitEventArgs args)
    {
        isInteracting = false;
        interactor = null;
    }

    /// <summary>
    /// This method is called when the actionTrigger is released, initiating the emergency shutdown scenario on nppClient.
    /// </summary>
    /// <param name="context"> passes event specific arguments</param>
    private void OnActionTriggered(InputAction.CallbackContext context)
    {
        if (nppClient != null && isInteracting)
        {
            Debug.Log("Setting Emergency Shutdown Scenario...");
            nppClient.StartCoroutine(nppClient.SetEmergencyShutdownScenario());
        }
    }
}